[Netatmo] Add units of measurement using QuantityType

lolodomo opened the pull request:
Fixes #3353 
 
Suppress unit from the channel description and the documentation
Fix examples for humidex channels in the documentation
Suppress hardcoded units from the French translation
Suppress hardcoded units from the German translation
Remove the channels BoilerOn BoilerOff Unit WindUnit and PressureUnit
Change channel DewpointDepression to QuantityType and temperature unit
Updated documentation with removed channels
Channel Sp_Temperature using QuantityType
Suppress min/max/step attributes from the channel type setpointTemp
Few corrections to satisfy code analysis

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

lolodomo commented:
@clinique for review

Remains to handle:

* new units m/s, Beaufort and Knot for wind speed
* new unit mmHg for pressure
* are your computations of humidex, heat index and dew point considering that the temperature could be provided in Fahrenheit degrees ?
* QuantityType for dewPoint channel ?
* QuantityType for rain1 and rain24 ?

lolodomo commented:
@clinique : unfortunately, it does not work at all as I was expecting. If I change the temperature unit with the Netatmo app to Fahrenheit, I correctly get Fahrenheit as temperazture unit inside the binding but the temperature values like for example dashboardData.getTemperature() are still returned in Celsius degrees !

lolodomo commented:
I already proposed a PR in ESH to add mmHg and knot units.
Beaufort is still missing.

clinique commented:
@lolodomo : regarding your questions : 
* Beaufort is more a scale than a unit type - but well, could maybe be handled also...
* Regarding dashboardData, it maybe a normal behaviour of netamo API to always return dashboarddata on their own default unit system
* As documented in WeatherUtils, getHumidex is designed to handle input temperature in °C. But we must keep in mind that Humidex is an index, not really a temperature. Same for Heat Index. 
So if you want to keep computations unchanged, we must consider transforming t°F=>t°C=>getHumidex()=>h°C=>h°F. 
* dewPoint is a temperature, so no pb.

lolodomo commented:
If they use their own units in the API, I hope this is documented and the same in all countries ? Or the used units are in the data with the values ?
That would mean we really don't need the unit data we currently get and use to set the unit channels. They are user choices for presentation in client app. But OH has its own unit selection mechanism.
I will check the netatmo API now.

lolodomo commented:
I should have looked at the API first ! They clearly mentioned the used units. So I have now to simplify my code. We finally don't care about the units returned by the API  corresponding to the user settings in the netatmo app.

lolodomo commented:
@clinique : simplified version online.

The missing units are outside the scope of our binding, they have to be added in ESH framework. I already proposed a fix to add knot and mmHg, Beaufort has to be studied.

As temperature is provided in Celsius degrees by the API, your code to compute humidex, heat index and dew point is OK.
I will change dew point to a QuantityType with temperature unit.
Regarding the heat index, it looks like it is a temperature too ( https://fr.wikipedia.org/wiki/Indice_de_chaleur ) so I will change it to a QuantityType with a temperature unit.
humidex is just an index so need for a unit.

Rain in the last hour has certainly to be moved to a QuantityType too.

lolodomo commented:
With the data provided by my external module (temp 13,4 °C humidity 80%), the computed heat index is very high, 51,4 °C. Are you sure that your computation is correct ?

lolodomo commented:
@wborn : you are welcome if you want to review ;)

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/ESH-INF/thing/channels.xml:
```diff
@@ -173,18 +172,18 @@
-		<item-type>Number</item-type>
+		<item-type>Number:Dimensionless</item-type>
 		<label>CO2</label>
 		<description>Air Quality in ppm</description>
```
Please remove `in ppm`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/ESH-INF/thing/channels.xml:
```diff
@@ -173,18 +172,18 @@
-		<item-type>Number</item-type>
+		<item-type>Number:Dimensionless</item-type>
 		<label>Noise</label>
 		<description>Current Noise Level in dB</description>
```
Please remove `in db`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/ESH-INF/thing/channels.xml:
```diff
@@ -203,11 +202,11 @@
-		<item-type>Number</item-type>
+		<item-type>Number:Pressure</item-type>
 		<label>Pressure</label>
 		<description>Current pressure in hPa</description>
```
Please remove `in hPa`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/ESH-INF/thing/channels.xml:
```diff
@@ -225,19 +224,19 @@
-		<item-type>Number</item-type>
+		<item-type>Number:Pressure</item-type>
 		<label>Abs Pressure</label>
 		<description>Absolute pressure in hPa</description>
```
Please remove `in hPa`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -109,31 +109,31 @@ Number Netatmo_Indoor_CO2 "CO2" <carbondioxide> { channel = "netatmo:NAMain:home
-| PressureUnit        | Number    | Unit used to report Pressure values (0 for mbar, 1 for inHg or 2 for mmHg) |
+| Channel ID          | Item Type            | Description                                              |
+|---------------------|----------------------|----------------------------------------------------------|
+| Co2                 | Number:Dimensionless | Air Quality in ppm                                       |
```
Please remove `in ppm`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -109,31 +109,31 @@ Number Netatmo_Indoor_CO2 "CO2" <carbondioxide> { channel = "netatmo:NAMain:home
+| Co2                 | Number:Dimensionless | Air Quality in ppm                                       |
+| Temperature         | Number:Temperature   | Current temperature                                      |
+| TempTrend           | String               | Temperature Evolution Trend (up, down, stable)           |
+| Noise               | Number:Dimensionless | Current Noise Level in dB                                |
```
Please remove `in dB`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -109,31 +109,31 @@ Number Netatmo_Indoor_CO2 "CO2" <carbondioxide> { channel = "netatmo:NAMain:home
+| Temperature         | Number:Temperature   | Current temperature                                      |
+| TempTrend           | String               | Temperature Evolution Trend (up, down, stable)           |
+| Noise               | Number:Dimensionless | Current Noise Level in dB                                |
+| Pressure            | Number:Pressure      | Current pressure in hPa                                  |
```
Please remove `in hPa`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -109,31 +109,31 @@ Number Netatmo_Indoor_CO2 "CO2" <carbondioxide> { channel = "netatmo:NAMain:home
+| Noise               | Number:Dimensionless | Current Noise Level in dB                                |
+| Pressure            | Number:Pressure      | Current pressure in hPa                                  |
+| PressTrend          | String               | Pressure evolution trend for last 12h (up, down, stable) |
+| AbsolutePressure    | Number:Pressure      | Absolute pressure in hPa                                 |
```
Please remove `in hPa`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -109,31 +109,31 @@ Number Netatmo_Indoor_CO2 "CO2" <carbondioxide> { channel = "netatmo:NAMain:home
+| Pressure            | Number:Pressure      | Current pressure in hPa                                  |
+| PressTrend          | String               | Pressure evolution trend for last 12h (up, down, stable) |
+| AbsolutePressure    | Number:Pressure      | Absolute pressure in hPa                                 |
+| Humidity            | Number:Dimensionless | Current humidity in %                                    |
```
Please remove `in %`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -148,24 +148,24 @@ Number Netatmo_Outdoor_Temperature "Temperature" { channel = "netatmo:NAModule1:
+|---------------------|----------------------|----------------------------------------------------------|
+| Temperature         | Number:Temperature   | Current temperature                                      |
+| TempTrend           | String               | Temperature Evolution Trend (up, down, stable)           |
+| Humidity            | Number:Dimensionless | Current humidity in %                                    |
```
Please remove `in %`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -180,25 +180,25 @@ Number Netatmo_Indoor2_Temperature "Temperature" { channel = "netatmo:NAModule4:
-| RfStatus            | Number    | Signal Strength (0 for no signal, 1 for weak, 2 for average, 3 for good or 4 for excellent) |
+| Channel ID          | Item Type            | Description                                              |
+|---------------------|----------------------|----------------------------------------------------------|
+| Co2                 | Number:Dimensionless | Air Quality in ppm                                       |
```
Please remove `in ppm`.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -180,25 +180,25 @@ Number Netatmo_Indoor2_Temperature "Temperature" { channel = "netatmo:NAModule4:
+| Co2                 | Number:Dimensionless | Air Quality in ppm                                       |
+| Temperature         | Number:Temperature   | Current temperature                                      |
+| TempTrend           | String               | Temperature Evolution Trend (up, down, stable)           |
+| Humidity            | Number:Dimensionless | Current humidity in %                                    |
```
Please remove `in %`.

There are more units left in the tables. Please remove the units throughout the complete file.

wborn commented on addons/binding/org.openhab.binding.netatmo/ESH-INF/thing/channels.xml:
```diff
@@ -173,18 +172,18 @@
-		<item-type>Number</item-type>
+		<item-type>Number:Dimensionless</item-type>
 		<label>Noise</label>
 		<description>Current Noise Level in dB</description>
```
Is there a reason why some of the descriptions still contain the unit? The README.md also contain channel descriptions with a certain unit. New or custom measurement systems may get added in the future which may cause unforeseen unit conversions.

wborn replied:
I see @cweitkamp  had the same remark. :-)

lolodomo replied:
JUst forget to change that. Wait few minutes...

lolodomo commented:
Review comments handled.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/ESH-INF/thing/channels.xml:
```diff
@@ -249,19 +248,19 @@
+		<state readOnly="true" pattern="%.1f %unit%" />
 	</channel-type>
 
 	<channel-type id="dewPointDepression" advanced="true">
```
What about this channel? Doesn't it contain a temperature, too?

cweitkamp replied:
Exactly. It is a delta between two temperature. Hence it is a temperature, too (see https://en.m.wikipedia.org/wiki/Dew_point_depression)

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -447,16 +447,16 @@ Bridge netatmo:netatmoapi:home "Netatmo API" [ clientId="*********", clientSecre
-Number Netatmo_Indoor_Humidity            "Humidity [%d %%]"               <humidity>         { channel = "netatmo:NAMain:home:inside:Humidity" }
+Number:Temperature Netatmo_Indoor_Temperature         "Temperature [%.1f %unit%]"          <temperature>      { channel = "netatmo:NAMain:home:inside:Temperature" }
+Number:Dimensionless Netatmo_Indoor_Humidity            "Humidity [%d %unit%]"               <humidity>         { channel = "netatmo:NAMain:home:inside:Humidity" }
 Number Netatmo_Indoor_Humidex             "Humidex [%.1f °C]"              <temperature_hot>  { channel = "netatmo:NAMain:home:inside:Humidex" }
```
If the humidex is a dimensionless quantity we should think about removing the `°C` in the example.

lolodomo replied:
Correct.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/README.md:
```diff
@@ -466,12 +466,12 @@ Number Netatmo_Indoor_WindUnit            "WindUnit [MAP(netatmo_windunit.map):%
 String Netatmo_Outdoor_TempTrend          "TempTrend [%s]"                 <line>             { channel = "netatmo:NAModule1:home:outside:TempTrend" }
-Number Netatmo_Outdoor_Humidity           "Humidity [%d %%]"               <humidity>         { channel = "netatmo:NAModule1:home:outside:Humidity" }
+Number:Dimensionless Netatmo_Outdoor_Humidity           "Humidity [%d %unit%]"               <humidity>         { channel = "netatmo:NAModule1:home:outside:Humidity" }
 Number Netatmo_Outdoor_Humidex            "Humidex [%.1f °C]"              <temperature_hot>  { channel = "netatmo:NAModule1:home:outside:Humidex" }
```
see above

lolodomo commented:
Regarding the setpoint temperature, you're right. As this is a channel with a command too, I am not sure how to handle that properly. Where can I find an example ?

lolodomo commented:
Regarding the channels BoilerOn and BoilerOff, they are apparently no more used by the binding !
@clinique: can you please confirm ? If that's correct, we just have to clean the code by suppressing them.

lolodomo commented:
And honestly, I am not sure there is now any usage having the channels containing the units (Unit, WindUnit and PressureUnit). They provide the user settings in the Netatmo app but not the user settings in OH.
I would suggest to suppress these channels too.

lolodomo commented:
As a summary, 4 points are remaining :
* QuantityType with temperature unit for the channel DewpointDepression or not  ?
* QuantityType with temperature unit for the channel Sp_Temperature
* Suppression or not of the channels BoilerOn and BoilerOff ?
* Suppression or not of the channels Unit, WindUnit and PressureUnit ?

wborn commented:
> Regarding the setpoint temperature, you're right. As this is a channel with a command too, I am not sure how to handle that properly. Where can I find an example ?

You can cast the command to a QuantityType as well. See the [NestThermostatHandler](https://github.com/openhab/openhab2-addons/blob/f3d759bf5d1a8982ef2d2dd2999aeb6c41af19bb/addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/handler/NestThermostatHandler.java#L154).

cweitkamp commented:
> Where can I find an example ?

Have a look at the Nest binding, the AVM FRITZ! binding or PR #3506 (MAX! binding). All of them handle a setpoint temperature. You have to convert the QuantityType value to the target unit required by the API.

clinique commented:
@lolodomo : yes, dewpoint depression is a temperature difference, so , a temperature.
It really seems that boilerOn and boilerOff is no more updated by Netatmo API. Never known (beside the idea the name can give) what it was excepted, so feel free to remove.
Completely agree on the fact to remove the channels containing the units now that now they are encapsulated by QuantityType, they have no more interest.

lolodomo commented:
@wborn @cweitkamp : The channel type `setpointTemp` has min / max / step values defined. If I convert this channel type to Number:Temperature, how can these values min/max be applicable for °C and °F units ?
Can I put the unit in the min and max values ? How ? `min="4.5°C" max="30°C"` for example ?

lolodomo commented:
Ok, I have now handled all remaining points.

For the last one, the Sp_Temperature channel, I am not sure if my changed code is ok, I am not able to test it and I don't know if the netatmo API expects particular values. As the set point temperature could now be provided in °F by the user and then converted in °C by the binding before using the API, this could be important or not. And there is the question about the values for the min and max attributes of this channel too.

If needed, I can squash all my commits.

lolodomo commented:
@wborn : ok, attributes are now removed for the setpointTemp channel type.

Finally I squashed all my changes after the initial commit in one unique second commit.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/NetatmoBindingConstants.java:
```diff
@@ -168,8 +164,7 @@
 
-    public static final Set<String> MEASURABLE_CHANNELS = Stream.of(CHANNEL_BOILER_ON, CHANNEL_BOILER_OFF)
-            .collect(Collectors.toSet());
+    public static final Set<String> MEASURABLE_CHANNELS = Collections.emptySet();
```
Is there a reason to keep this static final? If so this won't work: `Collections.emptySet()` returns an immutable set.

lolodomo replied:
I kept it in case we need it one day but ok I could suppress it.

lolodomo replied:
@clinique : this set is now empty. Is the class MeasurableChannels still necessary ?

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoDeviceHandler.java:
```diff
@@ -52,7 +51,7 @@
     @Nullable
     protected DEVICE device;
-    protected NAUserAdministrative userAdministrative;
+    // protected NAUserAdministrative userAdministrative;
```
Can this be removed?

lolodomo replied:
I kept it in case we need it one day but ok I could suppress it.

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/homecoach/NAHealthyHomeCoachHandler.java:
```diff
@@ -39,7 +38,7 @@ protected NAHealthyHomeCoach updateReadings() {
         NAHealthyHomeCoachDataBody homecoachDataBody = getBridgeHandler().getHomecoachDataBody(getId());
         if (homecoachDataBody != null) {
-            userAdministrative = homecoachDataBody.getUser().getAdministrative();
+            // userAdministrative = homecoachDataBody.getUser().getAdministrative();
```
Can this be removed?

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAMainHandler.java:
```diff
@@ -40,8 +40,7 @@ protected NAMain updateReadings() {
         if (stationDataBody != null) {
-            userAdministrative = stationDataBody.getUser().getAdministrative();
-
+            // userAdministrative = stationDataBody.getUser().getAdministrative();
```
Can this be removed?

cweitkamp commented on addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/thermostat/NAPlugHandler.java:
```diff
@@ -43,8 +43,7 @@ protected NAPlug updateReadings() {
         if (thermostatDataBody != null) {
-            userAdministrative = thermostatDataBody.getUser().getAdministrative();
-
+            // userAdministrative = thermostatDataBody.getUser().getAdministrative();
```
Can this be removed?

lolodomo commented:
I will take a look to the AVM FRITZ! binding to see how you compute the set point temperature.

lolodomo commented:
Ok, everything should be fine now.

wborn commented on addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/thermostat/NATherm1Handler.java:
```diff
@@ -143,13 +149,22 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                        } else {
+                            spTemp = new BigDecimal(command.toString()).setScale(1, RoundingMode.HALF_UP);
+                        }
+                        if (spTemp != null) {
```
It currently looks like a redundant null check. The `QuantityType.toUnit(..)` call could return null according to my compiler. So if you add an additional null check there, this check will no longer be redundant.

lolodomo replied:
Sorry, I don't understand what you expect. As you mentioned, the compiler tells me that a NULL check is required and so I aaded it.

wborn replied:
If we are talking about the same compiler warning there is still a null check missing:

![selection_003](https://user-images.githubusercontent.com/12213581/39446012-5158be9e-4cbd-11e8-92bd-3e46bcb5a862.png)

Currently a NPE occurs before it can even do the null check.

You can solve that with a case that looks like: 

```java
case CHANNEL_SETPOINT_TEMP: {
    BigDecimal spTemp = null;
    if (command instanceof QuantityType) {
        QuantityType<Temperature> quantity = ((QuantityType<Temperature>) command)
                .toUnit(API_TEMPERATURE_UNIT);
        if (quantity != null) {
            spTemp = quantity.toBigDecimal().setScale(1, RoundingMode.HALF_UP);
        }
    } else {
        spTemp = new BigDecimal(command.toString()).setScale(1, RoundingMode.HALF_UP);
    }
    if (spTemp != null) {
        // Switch the thermostat to manual mode on the desired setpoint for given duration
        Calendar cal = Calendar.getInstance();
        cal.add(Calendar.MINUTE, getSetPointDefaultDuration());
        getBridgeHandler().getThermostatApi().setthermpoint(getParentId(), getId(), "manual",
                (int) (cal.getTimeInMillis() / 1000), spTemp.floatValue());
        updateState(channelUID, new QuantityType<Temperature>(spTemp, API_TEMPERATURE_UNIT));
        requestParentRefresh();
    }
    break;
}
```

lolodomo commented:
That would be really cool if this PR could be merged quickly because I have free time this week and I am now waiting for this merge to fix another issue in this binding ;)

martinvw commented:
@wborn afaik we base the releasenotes on closed issues so if the description is good it will work out, we only add the breaking changes separately

lolodomo commented:
I added a comment about UoM in the relase notes.

wborn commented:
> I added a comment about UoM in the relase notes.

That's good! But several breaking changes are missing from it like the channels that no longer exist due to the introduction of UoM.

The breaking changes should also added to [update.lst](https://github.com/openhab/openhab-distro/blob/master/distributions/openhab/src/main/resources/bin/update.lst).

lolodomo commented:
Removed channels added.
